zebra-state/src/service/queued_blocks.rs

```diff
@@ -196,4 +196,80 @@ impl QueuedBlocks {
     pub fn utxo(&self, outpoint: &transparent::OutPoint) -> Option<transparent::Utxo> {
         self.known_utxos.get(outpoint).cloned()
     }
+
+    /// Clears known_utxos, by_parent, and by_height, then drains blocks.
+    /// Returns all key-value pairs of blocks as an iterator
+    pub fn drain(&mut self) -> Drain<'_, block::Hash, QueuedNonFinalized> {
+        self.known_utxos.clear();
+        self.known_utxos.shrink_to_fit();
+        self.by_parent.clear();
+        self.by_parent.shrink_to_fit();
+        self.by_height.clear();
+
+        self.blocks.drain()
+    }
+}
+
+#[derive(Debug, Default)]
+pub(crate) struct SentHashes {
+    /// A Vec of growable ring bufs to be used for pruning outdated hashes that are
+    /// inserted in the order of their height in a given batch.
+    sent_bufs: Vec<VecDeque<(block::Hash, block::Height)>>,
+
+    /// The growable ring buf to be used for pruning outdated hashes that are mostly
+    /// inserted in the order of their height for this batch of sent hashes
+    curr_buf: VecDeque<(block::Hash, block::Height)>,
+
+    /// Stores a set of hashes that have been sent to the block write task but
+    /// may not be in the finalized state yet.
+    set: HashSet<block::Hash>,
+}
+
+impl SentHashes {
+    /// Inserts the hash into the set of sent hashes, and pushes the (hash, height)
+    /// onto the current buf to be used for pruning outdated hashes.
+    ///
+    /// Assumes that hashes are added in the order of their height between `finish_batch` calls
+    /// for efficient pruning.
+    //
+    // Trades off some memory usage to avoid iterating over the entire set when non-finalized blocks
+    // are sent to the block write task, as would be the case with a HashMap.
+    pub fn add_with_height(&mut self, hash: block::Hash, height: block::Height) {
+        self.curr_buf.push_back((hash, height));
+        self.set.insert(hash);
+    }
+
+    /// Replaces `curr_buf` with an empty VecDeque and pushes the previous `curr_buf` to `sent_bufs`
+    pub fn finish_batch(&mut self) {
+        if !self.curr_buf.is_empty() {
+            self.sent_bufs.push(std::mem::take(&mut self.curr_buf));
+        }
+    }
+
+    /// Iterates over each buf in `sent_bufs`, removing hashes from SentHashes until reaching
+    /// the first hash with a height above the `height_bound`
+    ///
+    /// Assumes that hashes in each `sent_buf` will be added in order of their heights
+    /// in order to remove all hashes that are below `height_bound`.
+    pub fn prune_by_height(&mut self, height_bound: block::Height) {
+        self.sent_bufs.retain_mut(|buf| {
```

Comment:
We could accidentally misuse this API by calling `finish_batch()` after `prune_by_height()`.

```suggestion
    /// Iterates over each buf in `sent_bufs`, removing hashes from SentHashes until reaching
    /// the first hash with a height above the `height_bound`.
    ///
    /// Assumes that hashes in each `sent_buf` will be added in order of their heights
    /// in order to remove all hashes that are below `height_bound`.
    pub fn prune_by_height(&mut self, height_bound: block::Height) {
        self.finish_batch();
        
        self.sent_bufs.retain_mut(|buf| {
```